efi/shim_load_handler.go

```diff
@@ -0,0 +1,341 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package efi
+
+import (
+	"bytes"
+	"crypto/x509"
+	"errors"
+
+	efi "github.com/canonical/go-efilib"
+	"github.com/canonical/tcglog-parser"
+	"golang.org/x/xerrors"
+)
+
+// shimLoadHandler is an implementation of imageLoadHandler for shim.
+type shimLoadHandler struct {
+	Flags     shimFlags
+	VendorDb  *secureBootDB
+	SbatLevel shimSbatLevel
+}
+
+// shimLoadHandlerConstructor is used to construct a function with the same
+// signature as newShimLoadHandler which can be used to construct a custom
+// shimLoadHandler.
+type shimLoadHandlerConstructor struct {
+	version   *shimVersion
+	sbatLevel *shimSbatLevel
+}
+
+func newShimLoadHandlerConstructor() *shimLoadHandlerConstructor {
+	return new(shimLoadHandlerConstructor)
+}
+
+func (c *shimLoadHandlerConstructor) WithVersion(ver shimVersion) *shimLoadHandlerConstructor {
+	c.version = &ver
+	return c
+}
+
+func (c *shimLoadHandlerConstructor) WithSbatLevel(level shimSbatLevel) *shimLoadHandlerConstructor {
+	c.sbatLevel = &level
+	return c
+}
+
+func (c *shimLoadHandlerConstructor) New(rules imageRuleSet, image peImageHandle) (imageLoadHandler, error) {
```

Comment:
is this meant to implement newImageLoadHandlerFn ?

Reply:
It is.